### PR TITLE
perf(text2semantic): pass only active KV cache prefix to SDPA

### LINKED ISSUE #1310: Optimize native AR decoding by attending only to the active KV-cache prefix (1.97x compiled speedup on GV100)
### Self Checks

- [x] I have thoroughly reviewed the project documentation (installation, training, inference) but couldn't find any relevant information that meets my needs. [English](https://speech.fish.audio/) [中文](https://speech.fish.audio/zh/) [日本語](https://speech.fish.audio/ja/) [Portuguese (Brazil)](https://speech.fish.audio/pt/)
- [x] I have searched for existing issues [search for existing issues]([https://github.com/langgenius/dify/issues](https://github.com/fishaudio/fish-speech/issues)), including closed ones.
- [x] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/fishaudio/fish-speech/issues/515)).
- [x] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [x] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing? Tell us your story.

I am using the native PyTorch `text2semantic/inference.py` path on both Apple MPS and NVIDIA CUDA. While profiling S2-Pro autoregressive decoding, I found that each decode step attends over the **physical capacity** of the KV cache rather than only the prefix populated so far.

At upstream commit [`e5e2926`](https://github.com/fishaudio/fish-speech/commit/e5e292632cb11e7a27b2b7487f58f612bc101e13):

- When `input_pos` is present, [`forward_generate()`  builds an attention mask whose K dimension is `self.max_seq_len`]( https://github.com/fishaudio/fish-speech/blob/e5e292632cb11e7a27b2b7487f58f612bc101e13/fish_speech/models/text2semantic/llama.py#L435-L445).
- [`KVCache.update()` returns the full preallocated K/V tensors, which are then expanded with `repeat_interleave` before SDPA](https://github.com/fishaudio/fish-speech/blob/e5e292632cb11e7a27b2b7487f58f612bc101e13/fish_speech/models/text2semantic/llama.py#L908-L934).

Masking the unused tail preserves correctness, but it does not prevent the full-capacity K/V tensors from being expanded and passed to attention.

For S2-Pro, the model configuration has `max_seq_len=32768`, `n_head=32`, `n_local_heads=8`, `head_dim=128`, and 36 slow-transformer layers. My representative prompt contained 234 tokens and generation ended at roughly 400 context tokens, so only about 0.7%-1.3% of the physical cache was active.

At FP16, after expanding 8 local KV heads to 32 query heads, full-capacity K+V represents about 512 MiB of logical tensor data per layer, or about 18 GiB across 36 layers per semantic frame. This is an estimate of logical tensor work, not simultaneous memory residency or measured GPU bandwidth, but it explains why the work can scale with cache capacity instead of the current context length.

This is especially visible in native eager inference, but `torch.compile` does not eliminate all of the redundant work either.

### 2. What is your suggested solution?

Keep the KV cache physically preallocated at `max_seq_len`, but pass an explicit active length through the AR generation path and slice K/V **before** head expansion and SDPA:

1. Use `kv_len=T` for prompt prefill.
2. Use `kv_len=T+i+1` for decode step `i`.
3. Build the attention mask with `:kv_len`, not `:max_seq_len`.
4. After updating the physical KV cache, slice `k[:, :, :kv_len]` and `v[:, :, :kv_len]`.
5. Only then apply `repeat_interleave` and call SDPA.

The physical cache allocation and in-place update behavior remain unchanged. The optimization only reduces the logical K dimension consumed by attention.

I prepared a minimal implementation based directly on the upstream commit:

- Exact upstream baseline: [`benchmark/upstream-native-cuda`](https://github.com/quantumxiaol/fish-speech/tree/benchmark/upstream-native-cuda), commit [`e5e2926`](https://github.com/quantumxiaol/fish-speech/commit/e5e292632cb11e7a27b2b7487f58f612bc101e13)
- Active-KV branch: [`perf/active-kv-native`](https://github.com/quantumxiaol/fish-speech/tree/perf/active-kv-native), commit [`c4146e7`](https://github.com/quantumxiaol/fish-speech/commit/c4146e7e0145e615c76c34ab1d54829fde1a06ed)
- Minimal diff: [`e5e2926...c4146e7`](https://github.com/quantumxiaol/fish-speech/compare/e5e292632cb11e7a27b2b7487f58f612bc101e13...c4146e7e0145e615c76c34ab1d54829fde1a06ed)
- Correctness tests: [`tests/test_active_kv_attention.py`](https://github.com/quantumxiaol/fish-speech/blob/c4146e7e0145e615c76c34ab1d54829fde1a06ed/tests/test_active_kv_attention.py)

The implementation passes the Python position as an integer from the generation loop, avoiding `.item()` or another device synchronization in the production path.

### CUDA benchmark

Environment:

- GPU: NVIDIA Quadro GV100, 32 GB
- OS: Ubuntu x86_64
- NVIDIA driver: 525.105.17 (`nvidia-smi` reports CUDA compatibility 12.0)
- Python: 3.12
- PyTorch: 2.8.0+cu128
- Triton: 3.4.0
- Model: S2-Pro, FP16, native PyTorch inference
- Model `max_seq_len`: 32768
- Encoded prompt shape: `[11, 234]`
- Seed: 42
- `temperature=0.8`, `top_p=0.8`, `top_k=30`
- `max_new_tokens=1024`
- Four samples per mode

I used the same model, reference tokens, target text, sampling parameters, and seed on both branches. The first sample was treated as warm-up; for compiled runs it also contains compilation. The table reports the median of samples 1-3. Generated lengths differ slightly because small FP16/kernel differences can change a stochastic sampling trajectory, so throughput is normalized as generated tokens per second.

| Native inference mode | Full-capacity KV | Active-prefix KV | Speedup | Per-token latency | Peak CUDA memory reserved |
|---|---:|---:|---:|---:|---:|
| Eager | 2.73 tok/s | 12.57 tok/s | **4.60x** | 366.3 -> 79.6 ms | 17.33 -> 15.16 GB |
| `torch.compile` | 16.65 tok/s | 32.74 tok/s | **1.97x** | 60.1 -> 30.5 ms | 15.72 -> 15.16 GB |

Individual measurements:

```text
Eager, full KV:    2.71, 2.73, 2.73, 2.73 tok/s
Eager, active KV: 12.19, 12.57, 12.56, 12.58 tok/s

Compile, full KV:   4.01 [compile], 16.65, 16.64, 16.68 tok/s
Compile, active KV: 3.32 [compile], 32.69, 32.76, 32.74 tok/s
```

The existing log's `Bandwidth achieved` value is intentionally not used here because it is derived from model size multiplied by tokens/s; it is not a hardware DRAM-bandwidth measurement.

Cold compile times are also not compared because the runs shared persistent Inductor caches. With `TORCH_LOGS=recompiles`, the active-KV implementation showed one recompile when `kv_len` first changed from 235. It did **not** recompile once per token, and subsequent samples remained stable at approximately 32.7 tok/s. This one-time dynamic-shape compile behavior may still be worth improving for one-shot CLI use, while it is amortized in a long-running service.

### 3. Additional context or comments

PR #1306 identifies the related full-capacity KV-cache allocation and proposes lowering max_seq_len as a memory workaround. This proposal addresses a different part of the same path: it preserves the full cache capacity while preventing the unused tail from being expanded and passed to attention on every decode step.

Issue [#1168](https://github.com/fishaudio/fish-speech/issues/1168) contains a possibly related discussion about slow native inference and `torch.compile`. Its hardware, prompt, platform, and cold/warm compile conditions are not directly comparable with this benchmark, so I am not using its reported throughput as evidence here. This report isolates the full-capacity KV path with an exact upstream-commit A/B comparison.

I first found this while profiling the MPS backend. On an Apple M4 Pro with 48 GB unified memory, FP16, and `max_seq_len=4096`, the same active-prefix change improved the measured semantic generation rate from 4.08 to 8.31 frames/s (**2.04x**) for a 234-token prompt. This MPS result came from a service-level diagnostic path, so its absolute metric should not be compared directly with the CUDA CLI's tok/s. I include it only as secondary evidence that the optimization is backend-independent.

For output validation:

- All four representative CUDA outputs (full/active x eager/compiled) decoded successfully.
- I manually listened to all four WAV files and observed no audible regression.
- The files are not bit-identical, which is expected for stochastic FP16 inference when changing reduction shapes or compiled kernels.
- The unit test provides the stricter deterministic float32 equivalence check.

Current limitations of this benchmark:

- One CUDA architecture was tested (Volta GV100); Ampere/Ada/Hopper results would still be useful.
- The benchmark uses one target/reference pair and four samples per mode.
- No Nsight or `nvidia-smi dmon` trace was collected during generation, so I am not claiming a measured GPU-occupancy or hardware-bandwidth result.
- Cold compilation was excluded from the speed comparison.

The CUDA comparison is nevertheless narrowly controlled: `perf/active-kv-native` is based on the exact upstream baseline commit and changes only two inference source files plus focused tests.

I am interested in contributing this optimization and can prepare a clean PR if the approach is acceptable.

### 4. Can you help us with this feature?

- [x] I am interested in contributing to this feature.

### PR DESCRIPTION
**Pull request by quantumxiaol**
**Is this PR adding new feature or fix a BUG?**

Performance optimization for the native PyTorch autoregressive inference path.

### Problem

During cached generation, the physical KV cache is preallocated at
`max_seq_len`. At every decode step, the current implementation expands the
full-capacity K/V tensors with `repeat_interleave` and passes the full K/V
sequence to SDPA. For S2-Pro, this means expanding and passing a key sequence
length of 32768 to SDPA even when only a few hundred positions have been
populated.

This describes the tensors supplied to SDPA; it does not assume how a
particular SDPA backend handles the masked, unpopulated tail internally.

### Change

This change keeps the physical KV cache at full capacity but limits attention
to the active prefix:

- prompt prefill uses `kv_len=T`;
- decode step `i` uses `kv_len=T+i+1`;
- the causal mask is sliced to the active K length;
- cached K/V are sliced before head expansion and SDPA.

The regular full-sequence `forward()` path used for training does not use the
inference KV cache and is unchanged. The new slicing is gated by
`self.kv_cache is not None` and is exercised by cached `forward_generate()`
inference.

### Benchmark results

S2-Pro, FP16, native PyTorch CLI inference, prompt length 234,
`max_seq_len=32768`. Throughput is the median of three warm samples:

| Mode | Full-capacity KV | Active-prefix KV | Speedup | Peak CUDA memory reserved (GB) |
|---|---:|---:|---:|---:|
| Eager | 2.73 tok/s | 12.57 tok/s | **4.60x** | 17.33 -> 15.16 |
| `torch.compile` | 16.65 tok/s | 32.74 tok/s | **1.97x** | 15.72 -> 15.16 |

CUDA environment: Quadro GV100 32 GB, PyTorch 2.8.0+cu128, Triton 3.4.0.

Generated lengths differed slightly because FP16/kernel differences can change
stochastic sampling trajectories, so throughput is normalized by the number of
generated tokens.

The memory values are decimal GB, matching the existing log calculation
`torch.cuda.max_memory_reserved() / 1e9`.

The same change was also validated on Apple MPS (M4 Pro 48 GB, FP16,
`max_seq_len=4096`), where semantic generation improved from 4.08 to
8.31 frames/s (**2.04x**). The MPS service metric and CUDA CLI tok/s are not
directly comparable; only their within-device A/B ratios are reported.

Cold compile time is excluded from the throughput comparison. The existing
`Bandwidth achieved` log value is also excluded because it is derived from
model size and tok/s rather than measured hardware DRAM bandwidth.

### Correctness and compatibility validation

- Upstream pre-commit hooks pass for both modified files.
- A CPU shape check confirmed that SDPA receives only the active K/V prefix
  while the physical KV cache retains full capacity.
- A CPU float32 comparison confirmed that active-prefix attention matches
  full-cache attention with a masked tail (`rtol=1e-5`, `atol=1e-6`).
- CUDA eager and `torch.compile` generation completed successfully.
- With `TORCH_LOGS=recompiles`, the compiled active-KV path recompiled once
  when `kv_len` first changed, not once per token; warm throughput remained
  stable at approximately 32.7 tok/s.
- Four representative CUDA outputs (full/active x eager/compiled) were decoded
  and manually listened to; no audible regression was observed.
- MPS end-to-end generation and manual listening completed successfully.

The benchmark results above were collected from
[`c4146e7`](https://github.com/quantumxiaol/fish-speech/commit/c4146e7e0145e615c76c34ab1d54829fde1a06ed).
The current draft contains only the two production source changes; the
standalone correctness tests used for validation remain available in that
commit.

**Is this pull request related to any issue? If yes, please link the issue.**

Closes #1310

### CODE REVIEW

**Review by Whale-Dolphin — CHANGES_REQUESTED** [review]
The active-prefix optimization is effective. I verified the current head on an NVIDIA H200 with S2-Pro FP16 through the official text2semantic CLI: eager warm throughput improved from 9.22 to 19.73 tok/s (2.14x), and torch.compile warm throughput improved from 68.60 to 113.33 tok/s (1.65x).

Before merging, please retain regression tests in the repository for the behavior this PR changes:

1. Compare full-cache attention with active-prefix attention on a small deterministic fixture, with an explicit numerical tolerance and argmax assertion rather than requiring bitwise equality.
2. Cover kv_len boundary validation and the prefill/decode progression, including the first decode position and the maximum valid prefix.
3. Add an end-to-end autoregressive decode regression that exercises the real kv_len call chain and checks output shape, finite logits, and a defined token or distribution-level quality contract.

This is important because my FP16 comparison was not bitwise identical: the first five decode positions kept the same argmax, but the maximum absolute logit difference was 0.078125, and same-seed full generations diverged in 170 of 630 acoustic code elements after autoregressive propagation. This does not by itself indicate a quality regression, but the acceptable numerical and generation behavior needs to be encoded in tests so future SDPA or cache changes cannot silently break it.

**Comment by quantumxiaol** [comment]
Thanks for the independent H200 validation and the regression-test
requirements.

I added the requested regression coverage in commit
[`3263548`](https://github.com/quantumxiaol/fish-speech/commit/3263548ce8cc76ed879165062dfc0076d84acb6f),
under
[`tests/test_active_kv_attention.py`](https://github.com/quantumxiaol/fish-speech/blob/3263548ce8cc76ed879165062dfc0076d84acb6f/tests/test_active_kv_attention.py).
Five deterministic CPU tests cover:

1. **Active-prefix vs. full-cache attention**
   - Compares float32 outputs with `rtol=1e-5`, `atol=1e-6`, and a separate
     argmax assertion, without requiring bitwise equality.
   - Verifies that SDPA receives only the active K/V prefix while the physical
     cache retains full capacity. The observed maximum absolute difference was
     `5.96e-08`.

2. **`kv_len` validation and progression**
   - Covers valid boundaries `1` and `8`, invalid boundaries `0` and `9`,
     prompt prefill, the first and subsequent decode positions, and the maximum
     valid prefix.
   - The observed `(K, V, mask)` lengths were
     `[(3, 3, 3), (4, 4, 4), (8, 8, 8)]`; the end-to-end `kv_len` progression
     was `[3, 4, 5]`.

3. **End-to-end autoregressive regression**
   - Exercises the real
     `generate -> decode_n_tokens -> decode_one_token_ar -> forward_generate`
     path.
   - Checks output shape, finite slow/fast logits, probability and argmax
     agreement, and deterministic greedy-token equality.

The end-to-end fixture uses `top_k=1`, so exact token equality is its explicit
greedy-token contract; it does not assume stochastic FP16 CUDA generations are
bitwise identical.

Local result:

```text
Ran 5 tests
OK
```

The `pre-commit.ci` check also passes.
